Scripting: start with examples, highlight recommended commands and selectors, more

iamakulov opened the pull request:
TODO:

- [x] Add “Browser support” for `clearCache`

---

Heya,

### What this does

This PR restructures the Scripting doc as follows:
* adds formal sections about the comments syntax and the selectors syntax
* pushes examples to the top of the page (since that’s what most people are jumping to anyway)
* cross-links all sections where appropriate
* normalizes spacing in code blocks (eg spacing in this code snippet looks really weird today ↓) and formatting
   <img width="725" alt="Screen Shot 2022-02-03 at 21 56 15" src="https://user-images.githubusercontent.com/2953267/152410325-531a7f41-034a-44f0-92a2-f2fd99abcb1c.png">

The goal is to make it easier to scan through this doc and find what you need.

I’m comfortable if you decide to amend/reject some of the changes or all of them.

### Why

WebPageTest is great. I’m loving the tool, but it still can be frustrating when you’re trying to write a scripting command, and it just doesn’t work without any clear feedback. Today, I got frustrated by this, went to scripting docs, and ⌘F-ed for “debug” out of despair – only to realize a very useful paragraph about debugging has been sitting in the document this whole time:

> You won't get a lot of feedback as to why a script failed. For debugging purposes it is easiest to limit scripts to navigate and exec/execAndWait commands which can be debugged locally in a browser's dev tools.

What this tells us is nobody (including myself) ever _reads_ docs. From my experience, people tend to scan pages across headings and jump straight to code examples. I took the liberty of restructuring the Scripting doc to make it easier to scan across it.

iamakulov commented on src/scripting.md:
````diff
@@ -5,66 +5,179 @@ eleventyNavigation:
-you could identify it as **id=lgnId1** , **name=loginId** or **tabindex=1**
+See more [commands](#recommended-commands) and [examples](#sample-scripts) below.
+
+## Recommended Commands
+
+For historical reasons, if a script step fails, it does so silently. If something doesn’t work, unfortunately, you won’t get any feedback why. If you’re new to scripting, we recommend using the following commands only. It’s easier to debug them as you can run them manually yourself.
+
+#### `navigate`
+Navigates the browser to the provided URL and waits for it to complete.
+Browser Support: IE, Chrome, Firefox, Safari
+
+```markup
+usage:   navigate <url>
+example: navigate http://webmail.aol.com
+
+<url> - URL to provide the browser for navigation (same as you would enter into the address bar)
+```
+
+#### `exec`
+Executes JavaScript.
+Browser Support: IE, Chrome, Firefox
+```markup
+usage:   exec <JavaScript code>
+example: exec window.setInterval('window.scrollBy(0,600)', 1000);
+```
+
+#### `execAndWait`
+Executes JavaScript and waits for the browser to complete any activity generated from the action.
+Browser Support: IE, Chrome, Firefox
+```markup
+usage:   execAndWait <JavaScript code>
+example: execAndWait window.setInterval('window.scrollBy(0,600)', 1000);
+```
````
This expands the old

> You won't get a lot of feedback as to why a script failed. For debugging purposes it is easiest to limit scripts to navigate and exec/execAndWait commands which can be debugged locally in a browser's dev tools.

paragraph.

I’m highlighting this in a separate section becauseI feel this is very important and useful to know. I would’ve certainly benefited if I had been told earlier to use `exec` instead of `click` or `setValue` (which, sometimes, don’t work without any clear feedback).

iamakulov commented on src/scripting.md:
````diff
@@ -5,66 +5,179 @@ eleventyNavigation:
+
+## Full Reference
+
+### Comments
+
+A line that starts with  `// ` is ignored. Use this for comments.
+
+```markup
+// Example comment
+navigate not-a-comment.com
+```
````
I had no idea comments are allowed. Putting them in a separate section so that they’re easier to find when you’re scanning through headers.

tkadlec replied:
Good call!

iamakulov commented on src/scripting.md:
````diff
@@ -5,66 +5,179 @@ eleventyNavigation:
+navigate not-a-comment.com
+```
+
+### Selectors
````
I had no idea `innerText=` exists – until I edited this doc today. To make this easier to discover, I

* put selectors into a separate section
* linked to this section from every command that works with selectors.

iamakulov commented on src/scripting.md:
```diff
@@ -5,66 +5,179 @@ eleventyNavigation:
+
+and the following selectors will not:
+
+* ❌ `class=input input-email` (spaces inside attribute values are not supported)
+* ❌ `class="input input-email"` (not even if quoted)
```
Is this correct? I can’t manage to make my `setValue` script work so I can’t test this.

tkadlec replied:
It is!

iamakulov commented on src/scripting.md:
```diff
@@ -5,66 +5,179 @@ eleventyNavigation:
+
+Also, for a different button, this will not work either:
+
+* ❌ `innerText=log in` (spaces inside the text are not supported)
```
Is this correct?

iamakulov commented on src/scripting.md:
```diff
@@ -5,66 +5,179 @@ eleventyNavigation:
 
-// bring up the login screen
-navigate	http://webmail.aol.com
+#### `innerHtml=<html>`
```
Is it `innerHtml` or `innerHTML`? The old doc used `innerHtml` but the DOM API is `innerHTML` so I’m not sure that wasn’t a typo.

iamakulov commented on src/scripting.md:
```diff
@@ -5,66 +5,179 @@ eleventyNavigation:
+
+Also, for a different link, this will not work either:
+
+* ❌ `innerText=<span>Open Dashboard</span>` (spaces inside the HTML are not supported)
```
Is this correct?

iamakulov commented on src/scripting.md:
````diff
@@ -5,66 +5,179 @@ eleventyNavigation:
+
+```markup
+// Click the Dashboard link
+click innerHtml=<span>Dashboard</span>
````
Does the example look correct? I never used this selector (and my tests in WPT don’t seem to be working).

iamakulov commented on src/scripting.md:
```diff
@@ -98,514 +211,551 @@ input: setCookie  %ORIGIN% foo=bar
+
+For the difference between `sendClick` and `sendClickAndWait`, see [`click`](#click) and [`clickAndWait`](#clickandwait).
+
+#### `type` / `typeAndWait`
```
Note: this replaces “(AndWait)” with the full command name (“typeAndWait”). People may be ⌘F-ing for `typeAndWait`, and they won’t find the command if its name is never mentioned in full!

tkadlec replied:
👍

iamakulov commented on src/scripting.md:
```diff
@@ -98,514 +211,551 @@ input: setCookie  %ORIGIN% foo=bar
+<commands>
+endif
+
+example:
+if run 1
+  if cached 0
+  <do something for first view of first run>
+  endif
+else
+  <do something else for everything but first run>
+endif
```
I indented inner `if's – is this actually supported?

tkadlec replied:
Ha! Um...I'll be honest, I didn't even realize these commands existed. :) Looks like they're not hooked up in the agent so let's pull this for now until we put them in there.

iamakulov commented on src/scripting.md:
````diff
@@ -98,514 +211,551 @@ input: setCookie  %ORIGIN% foo=bar
 ```
 
-#### expireCache
+#### `expireCache`
````
Should we mention `clearCache`? https://github.com/WPO-Foundation/webpagetest/issues/186

tkadlec replied:
Probably, yes. While it's not at an ideal state, I know a few folks have mentioned it would be useful.

iamakulov commented on src/scripting.md:
````diff
@@ -98,514 +211,551 @@ input: setCookie  %ORIGIN% foo=bar
 <seconds to sleep> - An integer indicating how long to sleep.  The allowable range is 1-30.
 ```
 
+## Debugging
````
Adding a separate “Debugging” section because people may be searching for this.

tkadlec replied:
Good call.

iamakulov commented:
Removed the `if` section and added the `clearCache` one.

While we’re on it: how do I find where the command is implemented? (For example, in case I want to check the browser support.) E.g., I tried searching for `expireCache` in https://github.com/WPO-Foundation/webpagetest and https://github.com/WPO-Foundation/wptagent but there were zero results.

tkadlec commented:
All the scripting commands are going to be in the agent itself: https://github.com/WPO-Foundation/wptagent/

Having said that, woof this PR is exposing how badly we needed to update these docs (I mean, we knew we did but still....). Looks like `expireCache` is no more. https://github.com/WPO-Foundation/wptagent/blob/b2a77105966b818c682476cc4bb09ed93e848dab/internal/devtools_browser.py#L604

iamakulov commented:
@tkadlec Thanks for that agent link! I wasn’t sure if that’s the right one, or there’s another agent with its source code somewhere else 🙂

* I updated the PR to remove `expireCache` and a couple other unsupported commands.

* I also added supported browsers for `clearCache` (from what I gather, it’s supported in Chrome Desktop/Mobile + Safari on iOS):

   <img width="375" alt="Screen Shot 2022-03-14 at 12 14 06" src="https://user-images.githubusercontent.com/2953267/158131418-17e78851-9f4f-4d82-9b5b-4e764b3afc22.png">

* Question re: the `Selectors` section. Am I getting it right that `innerText` and `innerHtml` selectors aren’t supported? I can’t see any code implements them in 

   https://github.com/WPO-Foundation/wptagent/blob/8351db43e0d87ec46d008adf9c7dac5f078d783f/internal/webpagetest.py#L1122-L1148
   
   If they’re unsupported indeed, I’ll remove that section, and the PR should be done 🙂

tkadlec commented:
@iamakulov It does look like they're unsupported, yeah.

Thanks for all the hard work on this! It's looking _great_. I think once those references are out, we're good here.

iamakulov commented:
Alright, done!

tkadlec commented:
@iamakulov Whew. Thanks for this!! Appreciate all the hard work. It looks great. :)
